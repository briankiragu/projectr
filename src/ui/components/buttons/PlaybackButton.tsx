import { type Component } from 'solid-js';

const PlaybackButton: Component<{
  isEnabled: boolean;
  icon: string;
  text: string;
  title?: string;
  handler?: () => void;
}> = (props) => (
  <button
    type="button"
    title={props.title}
    class="flex h-10 w-10 items-center justify-center gap-2 rounded-full font-semibold transition-colors hover:bg-gray-600/70 hover:text-gray-50 lg:h-auto lg:w-auto lg:rounded-lg lg:px-4"
    classList={{
      'hover:bg-transparent disabled:text-gray-300 hover:cursor-not-allowed': !props.isEnabled,
    }}
    disabled={!props.isEnabled}
    onClick={props.handler}
  >
    <span class="material-symbols-outlined">{props.icon}</span>
    <span class="hidden lg:inline">{props.text}</span>
  </button>
)

export default PlaybackButton;
