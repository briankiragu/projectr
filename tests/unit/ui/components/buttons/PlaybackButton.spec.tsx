import PlaybackButton from "@components/buttons/PlaybackButton";
import { render, screen } from "@solidjs/testing-library";
import { userEvent } from "@testing-library/user-event";
import { describe, expect, test, vi } from "vitest";

describe("<PlaybackButton />", () => {
  // Define the component values.
  const icon = "arrow_back";
  const text = "Previous verse";
  const title = "ArrowLeft";
  const fn = vi.fn();

  // Setup user events.
  const user = userEvent.setup();

  test("it should render an icon, text and a title", () => {
    // Render the button onto the virtual DOM.
    render(() => (
      <PlaybackButton
        isEnabled={true}
        icon={icon}
        text={text}
        title={title}
        handler={fn}
      />
    ));

    // From the screen, get the rendered element.
    const el = screen.getByRole("button");

    // Run the assertions.
    expect(el.innerHTML).toMatchSnapshot();
    expect(el).toContainHTML(
      `<span class="material-symbols-outlined transition">${icon}</span>`
    );
    expect(el).toHaveAccessibleName(`${icon} ${text}`);
    expect(el).toHaveAttribute("title", title);
  });

  test("it should be disabled", () => {
    render(() => (
      <PlaybackButton
        isEnabled={false}
        icon={icon}
        text={text}
        title={title}
        handler={fn}
      />
    ));

    const el = screen.getByRole("button");
    expect(el).toBeDisabled();
  });

  test("it should perform a function when clicked", async () => {
    render(() => (
      <PlaybackButton
        isEnabled={true}
        icon={icon}
        text={text}
        title={title}
        handler={fn}
      />
    ));

    const el = screen.getByRole("button");
    await user.pointer({ keys: "[MouseLeft]", target: el });
    await user.pointer({ keys: "[MouseLeft]", target: el });

    expect(fn).toHaveBeenCalledTimes(2);
  });
});
