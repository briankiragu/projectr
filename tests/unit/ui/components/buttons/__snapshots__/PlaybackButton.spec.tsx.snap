// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<PlaybackButton /> > it should render an icon, text and a title 1`] = `"<span class="material-symbols-outlined transition">arrow_back</span><span class="hidden md:inline">Previous verse</span>"`;
